File: api-gateway/services/ratelimit.py
"""API Gateway — shared rate limiting.

Two separate concerns, deliberately kept apart:

  * Per source address, to slow an attacker working through a password list.
  * Per username, so a distributed attempt against ONE account is still
    stopped even though each request comes from a different address.

Both are needed. An address limit alone falls to a botnet; a username limit
alone lets someone spray one password across every account.

In-process and per-worker: with `--workers 4` an attacker gets four times the
stated budget. That is a real limitation, and the right fix is a shared store
(Redis) if this is ever exposed to the internet. It is still the difference
between thousands of guesses a minute and a handful.
"""

import logging
import time
from collections import defaultdict
from typing import Optional

logger = logging.getLogger("api_gateway.ratelimit")


class SlidingWindowLimiter:
    """Allow `limit` events per `window` seconds, per key."""

    def __init__(self, limit: int, window: int, name: str = "limiter"):
        self.limit = limit
        self.window = window
        self.name = name
        self._buckets: dict[str, list[float]] = defaultdict(list)

    def _prune(self, key: str, now: float) -> list[float]:
        bucket = self._buckets[key]
        bucket[:] = [t for t in bucket if now - t < self.window]
        # Keep the table from growing without bound as keys age out.
        if not bucket:
            self._buckets.pop(key, None)
            return []
        return bucket

    def allow(self, key: str) -> bool:
        """Record an attempt. False when the caller is over budget."""
        now = time.monotonic()
        bucket = self._prune(key, now)
        if len(bucket) >= self.limit:
            logger.warning(f"{self.name}: over limit for '{key}'")
            return False
        self._buckets[key].append(now)
        return True

    def retry_after(self, key: str) -> int:
        """Seconds until the oldest attempt in the window ages out."""
        bucket = self._buckets.get(key) or []
        if not bucket:
            return 0
        return max(1, int(self.window - (time.monotonic() - bucket[0])))

    def reset(self, key: str) -> None:
        """Clear a key. Called on a SUCCESSFUL login so a legitimate user who
        mistyped twice is not still counting against the limit."""
        self._buckets.pop(key, None)

File: crates/common/src/ratelimit.rs
//! Shared in-process rate limiting.
//!
//! Ported from `api-gateway/services/ratelimit.py`. This is the cheap,
//! per-worker first-line damping layer; the authoritative cross-worker login
//! throttle counts failures out of `audit_log` instead.

use std::collections::HashMap;
use std::sync::Mutex;
use std::time::{Duration, Instant};

/// Allow `limit` events per `window`, per key.
pub struct SlidingWindowLimiter {
    limit: usize,
    window: Duration,
    name: String,
    buckets: Mutex<HashMap<String, Vec<Instant>>>,
}

impl SlidingWindowLimiter {
    pub fn new(limit: usize, window: Duration, name: impl Into<String>) -> Self {
        Self {
            limit,
            window,
            name: name.into(),
            buckets: Mutex::new(HashMap::new()),
        }
    }

    /// Record an attempt. `false` when the caller is over budget.
    pub fn allow(&self, key: &str) -> bool {
        let now = Instant::now();
        let mut map = self.buckets.lock().unwrap();
        let bucket = map.entry(key.to_string()).or_default();
        bucket.retain(|t| now.duration_since(*t) < self.window);
        if bucket.len() >= self.limit {
            if bucket.is_empty() {
                map.remove(key);
            }
            drop(map);
            tracing::warn!("{}: over limit for '{key}'", self.name);
            return false;
        }
        bucket.push(now);
        true
    }

    /// Seconds until the oldest attempt in the window ages out.
    pub fn retry_after(&self, key: &str) -> u64 {
        let map = self.buckets.lock().unwrap();
        let Some(bucket) = map.get(key).filter(|b| !b.is_empty()) else {
            return 0;
        };
        let elapsed = bucket[0].elapsed().as_secs();
        self.window.as_secs().saturating_sub(elapsed).max(1)
    }

    /// Clear a key. Called on a successful login so a legitimate user who
    /// mistyped twice is not still counting against the limit.
    pub fn reset(&self, key: &str) {
        self.buckets.lock().unwrap().remove(key);
    }
}

impl Default for SlidingWindowLimiter {
    fn default() -> Self {
        Self::new(30, Duration::from_secs(60), "limiter")
    }
}
